File: milk/milk/page/milk_quality_dashboa/milk_quality_dashboa.py
import frappe
from collections import defaultdict
from math import sqrt
import json
from datetime import datetime

@frappe.whitelist()
def get_dashboard_data(filters=None):
    if isinstance(filters, str):
        filters = json.loads(filters or "{}")
    filters = filters or {}

    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    driver = filters.get("driver")
    village = filters.get("village")
    shift = filters.get("shift") or "both"     # morning | evening | both
    animal = filters.get("animal") or "both"   # cow | buffalo | both

    # Build conditions
    conditions = []
    values = {}

    if from_date and to_date:
        conditions.append("mq.date BETWEEN %(from_date)s AND %(to_date)s")
        values.update({"from_date": from_date, "to_date": to_date})

    if driver:
        conditions.append("mq.driver = %(driver)s")
        values["driver"] = driver

    if village:
        conditions.append("s.custom_villages = %(village)s")
        values["village"] = village

    if shift in ("morning", "evening"):
        # Fields morning, evening exist on Milk Quality
        if shift == "morning":
            conditions.append("mq.morning = 1")
        else:
            conditions.append("mq.evening = 1")

    if animal in ("cow", "buffalo"):
        field = "mq.cow" if animal == "cow" else "mq.buffalo"
        conditions.append(f"{field} = 1")

    where_clause = " AND ".join(conditions) if conditions else "1=1"

    # Fetch raw data
    rows = frappe.db.sql(f"""
        SELECT
            mq.date,
            smq.supplier,
            COALESCE(NULLIF(smq.water,''), 0) AS water,
            COALESCE(NULLIF(smq.protein,''), 0) AS protein,
            COALESCE(NULLIF(smq.density,''), 0) AS density,
            COALESCE(NULLIF(smq.hardness,''), 0) AS hardness,
            COALESCE(NULLIF(smq.pont,''), 0) AS pont
        FROM `tabMilk Quality` mq
        JOIN `tabSupplier Milk Quality` smq ON smq.parent = mq.name
        LEFT JOIN `tabSupplier` s ON s.name = smq.supplier
        WHERE {where_clause}
    """, values, as_dict=True)

    if not rows:
        return {
            "summary": {
                "total_suppliers": 0,
                "total_samples": 0,
                "avg_protein": 0,
                "avg_water": 0,
                "avg_density": 0,
                "avg_hardness": 0,
                "avg_pont": 0
            },
            "rankings": {
                "best": {"protein": [], "water": [], "density": [], "hardness": [], "pont": []},
                "worst": {"protein": [], "water": [], "density": [], "hardness": [], "pont": []}
            },
            "timeseries": [],
            "distributions": {"protein": [], "water": [], "density": [], "hardness": [], "pont": []},
            "correlations": {"matrix": {}},
            "outliers": {"protein": {"count":0}, "water": {"count":0}, "density": {"count":0}, "hardness": {"count":0}, "pont": {"count":0}}
        }

    metrics = ["protein", "water", "density", "hardness", "pont"]

    # Summary
    total_samples = len(rows)
    suppliers_set = set([r["supplier"] for r in rows if r.get("supplier")])
    total_suppliers = len(suppliers_set)

    def avg(field):
        vals = [to_float(r.get(field)) for r in rows if to_float(r.get(field)) is not None]
        return (sum(vals) / len(vals)) if vals else 0

    summary = {
        "total_suppliers": total_suppliers,
        "total_samples": total_samples,
        "avg_protein": avg("protein"),
        "avg_water": avg("water"),
        "avg_density": avg("density"),
        "avg_hardness": avg("hardness"),
        "avg_pont": avg("pont")
    }

    # Rankings: per supplier averages, best and worst 5
    supplier_stats = defaultdict(lambda: defaultdict(list))
    for r in rows:
        supplier = r.get("supplier") or "غير معروف"
        for m in metrics:
            v = to_float(r.get(m))
            if v is not None:
                supplier_stats[supplier][m].append(v)

    supplier_avg = []
    for supplier, data in supplier_stats.items():
        entry = {"supplier": supplier}
        for m in metrics:
            arr = data.get(m, [])
            entry[m] = sum(arr)/len(arr) if arr else None
        supplier_avg.append(entry)

    def top5(metric, reverse=True):
        # For "water", best means lower water (reverse=False).
        arr = [ {"supplier": e["supplier"], "value": e[metric]} for e in supplier_avg if e.get(metric) is not None ]
        if metric == "water":
            reverse = False  # lower is better
        arr.sort(key=lambda x: x["value"], reverse=reverse)
        return arr[:5]

    def bottom5(metric):
        # For "water", worst means higher water
        arr = [ {"supplier": e["supplier"], "value": e[metric]} for e in supplier_avg if e.get(metric) is not None ]
        if metric == "water":
            arr.sort(key=lambda x: x["value"], reverse=True)  # higher is worse
        else:
            arr.sort(key=lambda x: x["value"], reverse=False) # lower is worse
        return arr[:5]

    rankings = {
        "best":   { m: top5(m) for m in metrics },
        "worst":  { m: bottom5(m) for m in metrics }
    }

    # Time series daily averages
    daily = defaultdict(lambda: defaultdict(list))
    for r in rows:
        d = to_date_str(r.get("date"))
        for m in metrics:
            v = to_float(r.get(m))
            if v is not None:
                daily[d][m].append(v)
    timeseries = []
    for d in sorted(daily.keys()):
        item = {"date": d}
        for m in metrics:
            arr = daily[d][m]
            item[m] = sum(arr)/len(arr) if arr else 0
        timeseries.append(item)

    # Distributions (histograms): 10 buckets per metric
    distributions = {}
    for m in metrics:
        vals = [to_float(r.get(m)) for r in rows if to_float(r.get(m)) is not None]
        if not vals:
            distributions[m] = []
            continue
        mn, mx = min(vals), max(vals)
        if mn == mx:
            # single bucket
            distributions[m] = [{"label": f"{round(mn,2)}", "count": len(vals)}]
        else:
            buckets = histogram(vals, bins=10)
            distributions[m] = [{"label": f"{round(b['start'],2)}–{round(b['end'],2)}", "count": b["count"]} for b in buckets]

    # Correlations (Pearson)
    corr_matrix = {m: {} for m in metrics}
    for i, a in enumerate(metrics):
        for j, b in enumerate(metrics):
            if j < i:
                corr_matrix[a][b] = corr_matrix[b][a]
                continue
            pairs = []
            for r in rows:
                va = to_float(r.get(a))
                vb = to_float(r.get(b))
                if va is not None and vb is not None:
                    pairs.append((va, vb))
            if len(pairs) < 3:
                corr = 0
            else:
                xa, xb = zip(*pairs)
                corr = pearson(xa, xb)
            corr_matrix[a][b] = corr
    correlations = {"matrix": corr_matrix}

    # Outliers (IQR method)
    outliers = {}
    for m in metrics:
        vals = sorted([to_float(r.get(m)) for r in rows if to_float(r.get(m)) is not None])
        if len(vals) < 4:
            outliers[m] = {"count": 0}
            continue
        q1 = percentile(vals, 25)
        q3 = percentile(vals, 75)
        iqr = q3 - q1
        lower = q1 - 1.5 * iqr
        upper = q3 + 1.5 * iqr
        count = sum(1 for v in vals if v < lower or v > upper)
        outliers[m] = {"count": count, "lower": lower, "upper": upper}

    return {
        "summary": summary,
        "rankings": rankings,
        "timeseries": timeseries,
        "distributions": distributions,
        "correlations": correlations,
        "outliers": outliers
    }


def to_float(v):
    if v is None:
        return None
    try:
        return float(v)
    except Exception:
        return None

def to_date_str(v):
    if not v:
        return ""
    if isinstance(v, str):
        return v[:10]
    if isinstance(v, datetime):
        return v.strftime("%Y-%m-%d")
    return str(v)[:10]

def histogram(values, bins=10):
    mn, mx = min(values), max(values)
    width = (mx - mn) / bins if bins else 1
    if width == 0:
        return [{"start": mn, "end": mx, "count": len(values)}]
    counts = [0]*bins
    for v in values:
        if v == mx:
            idx = bins - 1
        else:
            idx = int((v - mn) / width)
            if idx < 0: idx = 0
            if idx >= bins: idx = bins - 1
        counts[idx] += 1
    out = []
    for i in range(bins):
        start = mn + i*width
        end = mn + (i+1)*width if i < bins-1 else mx
        out.append({"start": start, "end": end, "count": counts[i]})
    return out

def mean(arr):
    return sum(arr)/len(arr) if arr else 0

def pearson(a, b):
    n = len(a)
    if n < 2:
        return 0
    ma, mb = mean(a), mean(b)
    num = sum((x - ma)*(y - mb) for x, y in zip(a, b))
    da = sqrt(sum((x - ma)**2 for x in a))
    db = sqrt(sum((y - mb)**2 for y in b))
    if da == 0 or db == 0:
        return 0
    return num / (da * db)

def percentile(sorted_vals, p):
    if not sorted_vals:
        return None
    k = (len(sorted_vals)-1) * (p/100.0)
    f = int(k)
    c = f + 1
    if c >= len(sorted_vals):
        return float(sorted_vals[f])
    d0 = sorted_vals[f] * (c - k)
    d1 = sorted_vals[c] * (k - f)
    return float(d0 + d1)

File: milk/milk/page/milk_quality_dashboa/milk_quality_dashboa.js
frappe.pages['milk-quality-dashboa'].on_page_load = function(wrapper) {
    const page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'لوحة مؤشرات جودة اللبن',
        single_column: true
    });

    // Filters
    let filters = {
        from_date: frappe.datetime.month_start(),
        to_date: frappe.datetime.month_end(),
        driver: '',
        animal: 'buffalo'
    };

    page.add_field({
        label: 'من تاريخ',
        fieldtype: 'Date',
        fieldname: 'from_date',
        default: filters.from_date,
        change() { filters.from_date = this.value; load_dashboard(); }
    });
    page.add_field({
        label: 'إلى تاريخ',
        fieldtype: 'Date',
        fieldname: 'to_date',
        default: filters.to_date,
        change() { filters.to_date = this.value; load_dashboard(); }
    });
    page.add_field({
        label: 'السائق',
        fieldtype: 'Link',
        options: 'Driver',
        fieldname: 'driver',
        change() { filters.driver = this.value; load_dashboard(); }
    });
    page.add_field({
        label: 'النوع',
        fieldtype: 'Select',
        fieldname: 'animal',
        options: [
            {label:'جاموس', value:'buffalo'},
            {label:'أبقار', value:'cow'}
        ],
        default: 'buffalo',
        change() { filters.animal = this.value; load_dashboard(); }
    });

    const container = $(`<div class="milk-dashboard p-3"></div>`).appendTo(page.body);

    // Styles
    const style = document.createElement('style');
    style.textContent = `
        :root{
            --c-bg:#f7f9fc; --c-card:#fff; --c-border:rgba(15,23,42,.08);
            --c-text:#0f172a; --c-mute:#475569;
            --c-primary:#2563eb; --c-info:#0284c7; --c-success:#16a34a; --c-danger:#dc2626; --c-neutral:#64748b;
        }
        .milk-dashboard{ color:var(--c-text); background:var(--c-bg); font-family:"Noto Sans Arabic","Segoe UI",Tahoma,Arial,sans-serif; }
        .milk-dashboard .card{ border-radius:14px; border:1px solid var(--c-border); background:var(--c-card); }

        /* Period summary */
        .period-summary{ text-align:center; padding:14px 10px; }
        .period-summary .title{ font-size:1.2rem; font-weight:900; margin-bottom:6px; }
        .period-summary .rowline{ display:flex; gap:10px; flex-wrap:wrap; justify-content:center; color:var(--c-mute); font-weight:700; }
        .period-summary .chip{ background:#f1f5f9; border:1px solid var(--c-border); border-radius:999px; padding:6px 10px; }
        .period-summary .sep{ color:#94a3b8; }

        /* KPI mini-cards */
        .kpis{ padding:10px 12px 14px; }
        .kpis-title{ text-align:center; font-weight:900; margin-bottom:10px; }

        .kpi-rows{ display:flex; flex-direction:column; gap:12px; align-items:center; }
        .kpi-row{
            display:flex; gap:12px; flex-wrap:wrap; justify-content:center;
            width:100%;
        }
        .kpi-card{
            min-width:180px; max-width:220px; flex:0 1 200px;
            border:1px solid var(--c-border); border-radius:12px; background:#fff;
            padding:12px 12px; text-align:center;
            box-shadow: 0 1px 0 rgba(15,23,42,.02);
        }
        .kpi-val{ font-weight:900; font-size:1.14rem; line-height:1; }
        .kpi-label{ margin-top:6px; color:var(--c-mute); font-weight:700; font-size:.92rem; }

        /* On very small screens: stack */
        @media (max-width:520px){
            .kpi-card{ min-width:140px; flex:1 1 46%; }
        }

        /* Lists */
        .two-lists{ display:grid; grid-template-columns:1fr 1fr; gap:16px; }
        @media (max-width:992px){ .two-lists{ grid-template-columns:1fr; } }
        .list-card .list-group-item{ display:flex; align-items:center; justify-content:space-between; font-size:.98rem; }
        .idx-badge{ width:28px; height:28px; border-radius:8px; background:#e2e8f0; color:#0f172a; display:flex; align-items:center; justify-content:center; font-weight:800; }
        .list-left{ display:flex; align-items:center; gap:10px; min-width:0; }
        .name{ overflow:hidden; text-overflow:ellipsis; white-space:nowrap; max-width:52vw; }
        .value-badge{ color:#0f172a; background:#f1f5f9; border-radius:999px; padding:6px 10px; font-weight:800; border:1px solid var(--c-border); min-width:70px; text-align:center; }

        /* Titles and chart */
        .section-title{ font-weight:900; font-size:1.1rem; text-align:center; margin:6px 0 10px; }
        .section-sub{ text-align:center; color:var(--c-mute); margin-top:-4px; margin-bottom:8px; }
        .toggle-bar{ display:flex; gap:8px; align-items:center; flex-wrap:wrap; justify-content:center; }
        .toggle-bar .btn{ border:1px solid var(--c-border); background:#fff; color:#0f172a; padding:6px 10px; border-radius:10px; cursor:pointer; font-weight:700; }
        .toggle-bar .btn.active{ background:#e6f0ff; border-color:#bfdbfe; color:#1e3a8a; }
        .chart-card{ min-height:360px; }
    `;
    document.head.appendChild(style);

    // Utils
    function arDigits(str, enabled=false){
        if (!enabled) return str;
        const map = {'0':'٠','1':'١','2':'٢','3':'٣','4':'٤','5':'٥','6':'٦','7':'٧','8':'٨','9':'٩','.':'٫'};
        return String(str).replace(/[0-9.]/g, d => map[d] ?? d);
    }
    const useArabicDigits = false;
    function fmt(v, d=2){ if(v===null||v===undefined||isNaN(v)) return '-'; return arDigits(Number(v).toFixed(d), useArabicDigits); }

    const P = { primary:'#2563eb', info:'#0284c7', success:'#16a34a', danger:'#dc2626', neutral:'#64748b' };

    // Build KPI rows: first 3 centered in the first row, rest in the next row (centered)
    function kpiCards(summary){
        const items = [
            { label:'إجمالي الموردين', value: fmt(summary.total_suppliers,0)},
            { label:'إجمالي العينات',  value: fmt(summary.total_samples,0)},
            { label:'متوسط البروتين',  value: fmt(summary.avg_protein)},
            { label:'متوسط الماء',      value: fmt(summary.avg_water)},
            { label:'متوسط الكثافة',    value: fmt(summary.avg_density)},
            { label:'متوسط الصلابة',    value: fmt(summary.avg_hardness)},
            { label:'متوسط البونت',     value: fmt(summary.avg_pont)}
        ];
        const firstRow = items.slice(0, 3);
        const secondRow = items.slice(3);

        function renderRow(row){
            return `
                <div class="kpi-row">
                    ${row.map(it => `
                        <div class="kpi-card">
                            <div class="kpi-val">${it.value}</div>
                            <div class="kpi-label">${it.label}</div>
                        </div>
                    `).join('')}
                </div>
            `;
        }

        return `
            <div class="kpis">
                <div class="kpis-title">المجاميع والمتوسطات</div>
                <div class="kpi-rows">
                    ${renderRow(firstRow)}
                    ${secondRow.length ? renderRow(secondRow) : ''}
                </div>
            </div>
        `;
    }

    // List and chart helpers
    function listCard(title, items, colorVar = P.primary) {
        const li = (items && items.length)
            ? items.map((x, idx) => `
                <li class="list-group-item">
                    <div class="list-left">
                        <span class="idx-badge">${idx+1}</span>
                        <span class="name">${frappe.utils.escape_html(x.supplier)}</span>
                    </div>
                    <span class="value-badge" style="color:${colorVar}">${fmt(x.value)}</span>
                </li>
              `).join('')
            : `<li class="list-group-item text-muted">لا يوجد بيانات</li>`;
        return `
            <div class="card list-card h-100">
                <div class="card-header">${title}</div>
                <ul class="list-group list-group-flush">${li}</ul>
            </div>
        `;
    }
    function metricBlock(title, subtitle, bestList, worstList, color) {
        return `
            <div class="col-12 mb-3">
                <div class="card">
                    <div class="card-body">
                        <div class="section-title">${title}</div>
                        ${subtitle ? `<div class="section-sub">${subtitle}</div>` : ''}
                        <div class="two-lists mt-2">
                            ${listCard('أفضل 5 موردين', bestList, color)}
                            ${listCard('أسوأ 5 موردين', worstList, color)}
                        </div>
                    </div>
                </div>
            </div>
        `;
    }
    function chartCard(id, title, subtitle='', withToggle=true) {
        const sub = subtitle ? `<div class="section-sub">${subtitle}</div>` : '';
        const toggle = withToggle ? `
            <div class="toggle-bar mt-2">
                <span class="btn btn-ma active" data-target="${id}" data-mode="on">إظهار المتوسط المتحرك (7)</span>
                <span class="btn btn-ma-off" data-target="${id}" data-mode="off">إخفاء المتوسط المتحرك</span>
            </div>
        ` : '';
        return `
            <div class="col-12 mb-3">
                <div class="card chart-card">
                    <div class="card-body">
                        <div class="section-title">${title}</div>
                        ${sub}
                        ${toggle}
                        <div id="${id}" class="mt-2"></div>
                    </div>
                </div>
            </div>
        `;
    }

    // Charts
    let charts = {};
    function destroyCharts(){ Object.values(charts).forEach(ch=>{ try{ ch.destroy(); }catch(e){} }); charts = {}; }
    function movingAvg(arr, w){
        if(!Array.isArray(arr)||arr.length===0||w<=1) return new Array(arr.length).fill(null);
        const out=new Array(arr.length).fill(null); let sum=0;
        for(let i=0;i<arr.length;i++){ const v=Number(arr[i])||0; sum+=v; if(i>=w) sum-= (Number(arr[i-w])||0); if(i>=w-1) out[i]=sum/w; }
        return out;
    }

    function render(r){
        container.empty(); destroyCharts();
        if(!r.message){ container.html(`<p class="text-muted">لا يوجد بيانات</p>`); return; }

        const { summary, rankings, timeseries } = r.message;

        // Period summary
        const animalLabel = filters.animal === 'buffalo' ? 'جاموس' : 'أبقار';
        const fromStr = frappe.datetime.global_date_format ? frappe.datetime.str_to_user(filters.from_date) : filters.from_date;
        const toStr   = frappe.datetime.global_date_format ? frappe.datetime.str_to_user(filters.to_date)   : filters.to_date;
        const driverStr = filters.driver ? `السائق: ${frappe.utils.escape_html(filters.driver)}` : null;

        container.append(`
            <div class="card mb-3">
                <div class="period-summary">
                    <div class="title">ملخص الفترة (${animalLabel})</div>
                    <div class="rowline">
                        <span class="chip">من: ${fromStr}</span>
                        <span class="sep">—</span>
                        <span class="chip">إلى: ${toStr}</span>
                        ${driverStr ? `<span class="sep">—</span><span class="chip">${driverStr}</span>` : ''}
                        <span class="sep">—</span>
                        <span class="chip">النوع: ${animalLabel}</span>
                    </div>
                </div>
            </div>
        `);

        // KPIs (first 3 centered in first row, rest in next row)
        container.append(`<div class="card mb-3">${kpiCards(summary)}</div>`);

        // Timeseries
        const labels = (timeseries || []).map(x => x.date);
        const s = {
            protein:(timeseries||[]).map(x=>Number(x.protein||0)),
            water:(timeseries||[]).map(x=>Number(x.water||0)),
            density:(timeseries||[]).map(x=>Number(x.density||0)),
            hardness:(timeseries||[]).map(x=>Number(x.hardness||0)),
            pont:(timeseries||[]).map(x=>Number(x.pont||0)),
        };
        const ma = {
            protein: movingAvg(s.protein,7),
            water: movingAvg(s.water,7),
            density: movingAvg(s.density,7),
            hardness: movingAvg(s.hardness,7),
            pont: movingAvg(s.pont,7),
        };

        // Rankings
        function stableSort(items, metric, best=true) {
            const lowerIsBetter = (metric === 'water');
            const wantAsc = (best && lowerIsBetter) || (!best && !lowerIsBetter);
            const arr = (items || []).slice();
            arr.sort((a, b) => {
                const va = Number(a.value), vb = Number(b.value);
                if (!isFinite(va) && !isFinite(vb)) return 0;
                if (!isFinite(va)) return 1;
                if (!isFinite(vb)) return -1;
                if (va === vb) {
                    const sa = (a.supplier || '').toString();
                    const sb = (b.supplier || '').toString();
                    return sa.localeCompare(sb, 'ar');
                }
                return wantAsc ? (va - vb) : (vb - va);
            });
            return arr.slice(0, 5);
        }
        const best = {
            protein: stableSort(rankings.best.protein, 'protein', true),
            water:   stableSort(rankings.best.water, 'water', true),
            density: stableSort(rankings.best.density, 'density', true),
            hardness:stableSort(rankings.best.hardness, 'hardness', true),
            pont:    stableSort(rankings.best.pont, 'pont', true),
        };
        const worst = {
            protein: stableSort(rankings.worst.protein, 'protein', false),
            water:   stableSort(rankings.worst.water, 'water', false),
            density: stableSort(rankings.worst.density, 'density', false),
            hardness:stableSort(rankings.worst.hardness, 'hardness', false),
            pont:    stableSort(rankings.worst.pont, 'pont', false),
        };

        const blocks = $(`<div class="row"></div>`);
        blocks.append($(metricBlock('البروتين (أعلى أفضل)', 'أفضل وأسوأ الموردين حسب متوسط البروتين', best.protein, worst.protein, P.primary)));
        blocks.append($(metricBlock('الماء (أقل أفضل)', 'أفضل وأسوأ الموردين حسب متوسط الماء', best.water, worst.water, P.info)));
        blocks.append($(metricBlock('الكثافة (أعلى أفضل)', 'أفضل وأسوأ الموردين حسب متوسط الكثافة', best.density, worst.density, P.success)));
        blocks.append($(metricBlock('الصلابة (أعلى أفضل)', 'أفضل وأسوأ الموردين حسب متوسط الصلابة', best.hardness, worst.hardness, P.neutral)));
        blocks.append($(metricBlock('البونت (أعلى الأفضل)', 'أفضل وأسوأ الموردين حسب متوسط البونت', best.pont, worst.pont, P.danger)));
        container.append(blocks);

        // Trend chart
        const namesAr = { protein:'البروتين', water:'الماء', density:'الكثافة', hardness:'الصلابة', pont:'البونت' };
        const datasetsBase = [
            { name:namesAr.protein, values:s.protein },
            { name:namesAr.water, values:s.water },
            { name:namesAr.density, values:s.density },
            { name:namesAr.hardness, values:s.hardness },
            { name:namesAr.pont, values:s.pont },
        ];
        const datasetsMA = [
            { name:`${namesAr.protein} (متحرك 7)`, values:ma.protein },
            { name:`${namesAr.water} (متحرك 7)`, values:ma.water },
            { name:`${namesAr.density} (متحرك 7)`, values:ma.density },
            { name:`${namesAr.hardness} (متحرك 7)`, values:ma.hardness },
            { name:`${namesAr.pont} (متحرك 7)`, values:ma.pont },
        ];

        container.append($(chartCard('trendChart', 'الاتجاه اليومي للمتوسطات', `المتوسط اليومي لكل قياس خلال الفترة (${animalLabel})`, true)));

        if(charts.trend){ try{ charts.trend.destroy(); }catch(e){} }
        function buildTrend(showMA=true){
            const colorsBase = [P.primary, P.info, P.success, P.neutral, P.danger];
            const colorsMA   = ['#93c5fd','#67b3df','#7bdc9a','#cbd5e1','#fdcaca'];
            const ds = showMA ? datasetsBase.concat(datasetsMA) : datasetsBase;
            const cols = showMA ? colorsBase.concat(colorsMA) : colorsBase;
            charts.trend = new frappe.Chart('#trendChart', {
                data: { labels, datasets: ds },
                type: 'line',
                height: 360,
                colors: cols,
                axisOptions: { xAxisMode:'tick', yAxisMode:'tick', shortenYAxisNumbers:1 },
                lineOptions: { regionFill:0, heatline:0, dotSize:2, hideDots:0 }
            });
        }
        buildTrend(true);

        container.on('click', '.toggle-bar .btn', function(){
            const mode = this.getAttribute('data-mode');
            const group = $(this).closest('.toggle-bar');
            group.find('.btn').removeClass('active');
            $(this).addClass('active');
            buildTrend(mode === 'on');
        });
    }

    function load_dashboard(){
        frappe.call({
            method: "milk.milk.page.milk_quality_dashboa.milk_quality_dashboa.get_dashboard_data",
            args: { filters },
            callback: function(r){ render(r); }
        });
    }

    load_dashboard();
};